### PR TITLE
Log Windows product type and installation type, and Linux libc version

### PR DESCRIPTION
**Pull request by danmosemsft**
Datapoints and rationale are described in the previous PR https://github.com/dotnet/cli/pull/8579

### CODE REVIEW

File: src/dotnet/Telemetry/ExternalTelemetryProperties.cs [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+            // Catch everything: this is for telemetry only
+            catch(Exception e)
+            {
+                Debug.Assert(false, $"Unexpected exception from GetProductInfo: ${e.GetType().Name}: ${e.Message}");
+            }
+
+            return "";
+        }
+
+        [DllImport("libc", ExactSpelling = true, CharSet = CharSet.Ansi, CallingConvention = CallingConvention.Cdecl)]
```
Comment (nguerrera) [inline src/dotnet/Telemetry/ExternalTelemetryProperties.cs]: Is CharSet needed without any string marshaling here?

File: src/dotnet/Telemetry/ExternalTelemetryProperties.cs [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+        private static extern IntPtr gnu_get_libc_release();
+
+        [DllImport("libc", ExactSpelling = true, CharSet = CharSet.Ansi, CallingConvention = CallingConvention.Cdecl)]
+        private static extern IntPtr gnu_get_libc_version();
+
+        /// <summary>
+        /// If gnulibc is available, returns the release, such as "stable".
+        /// If the libc is musl, currently returns empty string.
```
Comment (eerhardt) [inline src/dotnet/Telemetry/ExternalTelemetryProperties.cs]: Is this not possible?  or is it just a lower priority?

File: test/dotnet.Tests/TelemetryCommonPropertiesTests.cs [hunk 3]
```diff
@@ -52,6 +52,61 @@ public void TelemetryCommonPropertiesShouldContainKernelVersion()
             unitUnderTest.GetTelemetryCommonProperties()["Kernel Version"].Should().Be(RuntimeInformation.OSDescription);
         }
 
+        [WindowsOnlyFact]
+        public void TelemetryCommonPropertiesShouldContainWindowsInstallType()
+        {
+            var unitUnderTest = new TelemetryCommonProperties(getMACAddress: () => null, userLevelCacheWriter: new NothingCache());
+            var assignedMachineId = unitUnderTest.GetTelemetryCommonProperties()["Installation Type"].Should().NotBeEmpty();
```
Comment (eerhardt) [inline test/dotnet.Tests/TelemetryCommonPropertiesTests.cs]: I assume `assignedMachineId` is a copy-paste left over.

File: src/dotnet/Telemetry/ExternalTelemetryProperties.cs (hunk 2 above)
Comment (danmosemsft) [reply src/dotnet/Telemetry/ExternalTelemetryProperties.cs]: Dug through Musl briefly and did not find a equivalent. As for version it does not seem to be exposed by an API.

File: src/dotnet/Telemetry/ExternalTelemetryProperties.cs [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+using System.Runtime.InteropServices;
+using System.Diagnostics;
+
+namespace Microsoft.DotNet.Cli.Telemetry
+{
+    // Some properties we need for telemetry, that don't yet have suitable
+    // public API
+    internal static class ExternalTelemetryProperties
+    {
+        /// For Windows, returns the OS installation type, eg. "Nano Server", "Server Core", "Server", or "Client".
```
Comment (eerhardt) [inline src/dotnet/Telemetry/ExternalTelemetryProperties.cs]: (nit) missing an opening `<summary>` to match the close below.

**Comment by danmosemsft** [comment]
@wli3 could you please revew, and merge if you're comfortable?

**Comment by danmosemsft** [comment]
```
[xUnit.net 00:11:34.2523151] D:\j\workspace\debug_windows---8a02d2bd\bin\2\win81-x86\dotnet\sdk\2.1.300-preview2-008274\NuGet.targets(114,5): error : The feed 'cli-deps [https://dotnet.myget.org/F/cli-deps/api/v3/index.json]' lists package 'System.ComponentModel.EventBasedAsync.4.0.11' but multiple attempts to download the nupkg have failed. The feed is either invalid or required packages were removed while the current operation was in progress. Verify the package exists on the feed and try again. [D:\j\workspace\debug_windows---8a02d2bd\bin\3\win81-x86\test\dotnet-test.Tests\8\VSTestCore\VSTestCore.csproj]
Indication 1 
```
@dotnet-bot test Windows_NT x86 Debug Build

**Comment by danmosemsft** [comment]
@dotnet-bot test Ubuntu x64 Release Build  please (`[39;49mBuild timed out (after 120 minutes). Marking the build as aborted.`)

This leg normally takes 30-40 mins

**Comment by wli3** [comment]
@dotnet-bot test Ubuntu x64 Release Build

**Comment by wli3** [comment]
I will drive the review

**Comment by wli3** [comment]
Speaking perf, how long does it take for the thread to finish? What is the percentage added compared with before?

**Comment by wli3** [comment]
There is a network call at the end of the thread. As long as it is not slower than that, it should be fine.

File: src/dotnet/Telemetry/ExternalTelemetryProperties.cs [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+        internal static string GetProductType()
+        {
+            if (RuntimeEnvironment.OperatingSystemPlatform != Platform.Windows)
+                return "";
+
+            try
+            {
+                if (GetProductInfo((uint)Environment.OSVersion.Version.Major, (uint)Environment.OSVersion.Version.Minor, 0, 0, out uint productType))
+                {
+                    return productType.ToString("D");
```
Comment (nguerrera) [inline src/dotnet/Telemetry/ExternalTelemetryProperties.cs]: Pass CultureInfo.InvariantCulture here.  I don't think the "D" format actually varies by culture (does it?) but better not to have to think about the risk.

Comment (danmosemsft) [reply src/dotnet/Telemetry/ExternalTelemetryProperties.cs]: In a quick look at the code it apparently doesn't (at least for uint) but I'll add it.

**Comment by danmosemsft** [comment]
> Speaking perf, how long does it take for the thread to finish? What is the percentage added compared with before?

I logged the time in Windows and Linux, the cost is a few single milliseconds. Most of it is in existing code.

**Comment by wli3** [comment]
Thanks, that's great. I also see Nick's code review is addressed . I'll merge it once CI is done

**Pull request merged**